executor/exeutor_test.go

```diff
@@ -0,0 +1,101 @@
+// Copyright 2015 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package executor_test
+
+import (
+	"testing"
+
+	. "github.com/pingcap/check"
+	"github.com/pingcap/tidb"
+	"github.com/pingcap/tidb/executor"
+	"github.com/pingcap/tidb/kv"
+	"github.com/pingcap/tidb/util/testkit"
+)
+
+func TestT(t *testing.T) {
+	TestingT(t)
+}
+
+var _ = Suite(&testSuite{})
+
+type testSuite struct {
+	store kv.Storage
+}
+
+func (s *testSuite) SetUpSuite(c *C) {
+	store, err := tidb.NewStore("memory://test/test")
+	c.Assert(err, IsNil)
+	s.store = store
+}
+
+func (s *testSuite) TearDownSuite(c *C) {
+	s.store.Close()
+}
+
+func (s *testSuite) TestPrepared(c *C) {
+	tk := testkit.NewTestKit(c, s.store)
+	tk.MustExec("use test")
+	tk.MustExec("drop table if exists prepare_test")
+	tk.MustExec("create table prepare_test (id int PRIMARY KEY AUTO_INCREMENT, c1 int, c2 int, c3 int default 1)")
+	tk.MustExec("insert prepare_test (c1) values (1),(2),(NULL)")
+
+	tk.MustExec(`prepare stmt_test_1 from 'select id from prepare_test where id > ?'; set @a = 1; execute stmt_test_1 using @a;`)
+	tk.MustExec(`prepare stmt_test_2 from 'select 1'`)
+	// Prepare multiple statement is not allowed.
+	_, err := tk.Exec(`prepare stmt_test_3 from 'select id from prepare_test where id > ?;select id from prepare_test where id > ?;'`)
+	c.Assert(executor.ErrPrepareMulti.Equal(err), IsTrue)
+	// The variable count do not match.
+	_, err = tk.Exec(`prepare stmt_test_4 from 'select id from prepare_test where id > ? and id < ?'; set @a = 1; execute stmt_test_4 using @a;`)
+	c.Assert(executor.ErrWrongParamCount.Equal(err), IsTrue)
```

Comment:
I thing invalid is all about the count itself, but wrong means it is different than the count of the prepared statement.